lib/nostr.js

```diff
@@ -17,155 +16,151 @@ export const DEFAULT_CROSSPOSTING_RELAYS = [
 
-        if (accepted) {
-          resolve(eventId)
-        } else {
-          reject(new Error(reason || `event rejected: ${eventId}`))
-        }
-      }
+    const successfulRelays = []
+    const failedRelays = []
 
-      ws.addEventListener('message', listener)
+    const relaySet = NDKRelaySet.fromRelayUrls(relays, this.ndk, true)
 
-      ws.send(JSON.stringify(['EVENT', event]))
+    relaySet.on('publish:failed', (relay, error) => {
+      failedRelays.push({ relay, error })
     })
 
-    try {
-      return await withTimeout(ackPromise, timeout)
-    } finally {
-      ws.removeEventListener('message', listener)
+    successfulRelays.push(...await relaySet.publish(event, timeout))
+
+    return {
+      event,
+      successfulRelays,
+      failedRelays
     }
   }
 
-  async fetch (filter, { timeout } = {}) {
-    const ws = this.ws
-
-    let listener
-    const ackPromise = new Promise((resolve, reject) => {
-      const id = crypto.randomBytes(16).toString('hex')
-
-      const events = []
-      let eose = false
-
-      listener = function onmessage (msg) {
-        const [type, subId, event] = JSON.parse(msg.data)
-
-        if (subId !== id) return
-
-        if (type === 'EVENT') {
-          events.push(event)
-          if (eose) {
-            // EOSE was already received:
-            // return first event after EOSE
-            resolve(events)
-          }
-          return
-        }
-
-        if (type === 'CLOSED') {
-          return resolve(events)
-        }
-
-        if (type === 'EOSE') {
-          eose = true
-          if (events.length > 0) {
-            // we already received events before EOSE:
-            // return all events before EOSE
-            ws.send(JSON.stringify(['CLOSE', id]))
-            return resolve(events)
-          }
-        }
+  /* eslint-disable camelcase */
+  async crosspost ({ created_at, content, tags = [], kind }, { relays = DEFAULT_CROSSPOSTING_RELAYS, privKey, signer, timeout } = {}) {
+    try {
+      const { event: signedEvent, successfulRelays, failedRelays } = await this.publish({ created_at, content, tags, kind }, { relays, privKey, signer, timeout, signWithExtension: true })
+
+      let noteId = null
+      if (signedEvent.kind !== 1) {
+        noteId = await nip19.naddrEncode({
+          kind: signedEvent.kind,
+          pubkey: signedEvent.pubkey,
+          identifier: signedEvent.tags[0][1]
+        })
+      } else {
+        noteId = hexToBech32(signedEvent.id, 'note')
       }
 
-      ws.addEventListener('message', listener)
-
-      ws.send(JSON.stringify(['REQ', id, ...filter]))
-    })
-
-    try {
-      return await withTimeout(ackPromise, timeout)
-    } finally {
-      ws.removeEventListener('message', listener)
+      return { successfulRelays, failedRelays, noteId }
+    } catch (error) {
+      console.error('Crosspost error:', error)
+      return { error }
     }
   }
 }
 
+/**
+ * @type {Nostr}
+ */
+export default new Nostr()
```

Comment:
this is the global singleton, connections to specific relays are handled with RelaySets and their life cycle is left to ndk to handle